Implement Heroku Teams commands + introduces invitations

raulb opened the pull request:
# The Heroku CLI welcomes Heroku Teams
## New commands
- `heroku teams` - List only your Heroku Teams
## Previous commands that have been modified
- `heroku orgs` will show a message at the end mentioning that if you have teams, you can use `heroku teams` directly
- `heroku orgs` now includes two new flags: `--enterprise` or `--teams`
  - `heroku orgs --enterprise` will show you only your Heroku Enterprise Organizations
  - `heroku orgs --teams` will show you only your Heroku Teams (this is equivalent to `heroku teams`).
- `heroku members:add`, `heroku members:remove` and `heroku members` accept the flag `--team` or `-t` to add, remove, or list members from your team. If you use `-o` or `--org` in a Heroku Team, it'll work as usual, but a warning will be shown at the end of the output.

_Example:_

```
$ heroku members -o team-name
user@heroku.com         member
user+member@heroku.com  member
user-admin@heroku.com   admin
 ▸    team-name is a Heroku Team
 ▸    Heroku CLI now supports Heroku Teams.
 ▸    Use -t or --team for teams like team-name
```
## Introducing Heroku Invites
- `heroku members -t TEAM` will include not only existing team members, but also pending invitations. 

_Example:_

```
$ heroku members -t team-name
new-user@email.com             role                 pending 
user@email.com                 role
user@email.com                 role
```
- `heroku members` now includes the flag `--pending` which it'll list only pending invitations in your team:

```
$ heroku members -t team-name --pending
new-user@email.com             role                 pending 
```
- You'll be able to revoke an invite (in those members who haven't accepted yet) by the known command `heroku members:remove -t TEAM`.

_Example:_

```
$ heroku members:remove user@email.com -t team
Revoking invite for user@email.com in team... done
```
- Removing existing Heroku Team or Heroku Organization members will work as usual:

```
$ heroku members:remove user@email.com -t team
Removing user@email.com from team... done
```
### Other changes
- Warning regarding team limits will be dependent on team invitations apart from existing team members. 

_Example (assuming there were 5 members):_

```
$ heroku members:add user@email.com -r admin -t TEAM
Inviting user@email.com to my-team as role... email sent
 ▸    You'll be billed monthly for teams over 5 members.
```
## Already launched as part of #66
- You can invite members to your team by using `heroku members:add`. 
  If the team has been flagged in with `team-invite-acceptance`, adding members will trigger an invitation in form of an email with the following output:

```
 $ heroku members:add user@email.com -t my-team --role ROLE
 Inviting user@email.com to my-team as role... email sent
```

raulb committed:
Implement --teams and invitations

raulb commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;2VDFgXx8&#x2F;383-update-cli-to-list-create-invitations-l"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Update CLI to list&#x2F;create invitations [L]</a>

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
   topic: 'members',
   command: 'add',
-  description: 'adds a user to an organization',
+  description: 'adds an user to an organization or a team',
```
'adds a user to an organization or a team' (not 'an user')

raulb replied:
fixed, thanks

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
-  if ((members.length === (FREE_TEAM_LIMIT + 1)) && orgCreationFeature) {
-    cli.log("You'll be billed monthly for teams over 5 members.")
+
+  if (orgInfo.type === 'team') {
```
I'm personally not a big fan of having this pinned to a team. Why aren't we only checking if the org is flagged in or not?

raulb replied:
See https://github.com/heroku/heroku-orgs/pull/66#discussion_r80800844

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
+  let orgInfo = yield Utils.orgInfo(context, heroku)
+  let groupName = context.org || context.flags.team
+
+  let warnMembershipLimit = function (totalMembers) {
```
Should the function variables be `const` instead of `let`? There's no need for them to mutate.

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
+    let teamMembers = 0
+    let teamInvites = 0
+
+    teamInvites = heroku.request({
```
there's a weird type issue here. `teamInvite` and `teamMembers` are initialized with `0`, but then transformed into promises of JSON responses. Should they instead be initialized with empty promises or null?

raulb replied:
Fixed by https://github.com/heroku/heroku-orgs/pull/66.

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
-  const FREE_TEAM_LIMIT = 5
-  let org = context.org
+  let orgInfo = yield Utils.orgInfo(context, heroku)
+  let groupName = context.org || context.flags.team
```
There's mixed terminology between orgs, teams, and groups. I'd suggest picking one and using it consistently.

raulb replied:
Fixed by https://github.com/heroku/heroku-orgs/pull/66

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
+
+    let orgFeatures = yield heroku.get(`/organizations/${groupName}/features`)
+
+    if (orgFeatures.filter(of => of.name === 'team-invite-acceptance').length) {
```
`includes()` would be clearer here.

raulb replied:
Changed, thanks

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
+    }
+
+    Promise.all([teamInvites, teamMembers]).then(membersAndInvites => {
+      let membersCount = membersAndInvites.reduce((m, i) => {
```
Rather than `reduce`, destructuring might make this clearer.

``` javascript
let [members, invites] = membersAndInvites;
return members.length + invites.length;
```

raulb replied:
Changed by https://github.com/heroku/heroku-orgs/pull/66.

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
+
+  let warnMembershipLimit = function (totalMembers) {
+    // Users receive `You'll be billed monthly for teams over 5 members.`
+    if (totalMembers === 6) {
```
Ouch...is there a way to avoid hard-coding this? At the least have a constant somewhere in the code?

raulb replied:
Changed by https://github.com/heroku/heroku-orgs/pull/66

naaman commented on commands/members/add.js:
```diff
@@ -3,46 +3,96 @@
+      warnMembershipLimit(membersCount)
+    })
+  } else {
+    yield addMemberToOrg(email, role, context.org)
```
This will fail if the org is `enterprise` and flagged into invitations.

raulb replied:
Changed by https://github.com/heroku/heroku-orgs/pull/66.

naaman commented on commands/members/index.js:
```diff
@@ -1,37 +1,67 @@
+  let groupName = context.org || context.flags.team
+  let teamInvites = []
+
+  if (orgInfo.type === 'team') {
```
Same here. We should use the flag.

raulb replied:
https://github.com/heroku/heroku-orgs/pull/66#discussion_r80800844

naaman commented on commands/members/index.js:
```diff
@@ -1,37 +1,67 @@
+  if (context.flags.pending) members = members.filter(m => m.status === 'pending')
   if (context.flags.json) {
-    cli.log(JSON.stringify(members, null, 2))
+    cli.log(JSON.stringify(members, null, 3))
```
This seems a bit disingenuous. Isn't the `json` flag a representation of the API? I realize we have two types of things here, but perhaps we should serialize them as those two things rather than the union of the two of them.

naaman commented on commands/members/remove.js:
```diff
@@ -2,24 +2,69 @@
-    path: `/organizations/${org}/members/${encodeURIComponent(email)}`
-  })
-  yield cli.action(`Removing ${cli.color.cyan(email)} from ${cli.color.magenta(org)}`, request)
+  let teamInvites = function * () {
```
This is repeated a few times. Can you move this into a helper, utility, or model to reduce the repetition?

naaman replied:
And, FWIW, it probably makes sense to move all the API calls into a helper, utility, or model.

raulb replied:
Ideally, without the use of variants, the api call would be just `heroku.get(path)` which is no different than using a helper, etc. This should be much cleaner once the endpoint is in v3.

eablack commented on commands/access/add.js:
```diff
@@ -20,7 +20,7 @@ function * run (context, heroku) {
   }
 
-  if (_.includes(_.map(orgFeatures, 'name'), 'org-access-controls')) {
+  if (orgFeatures.filter(of => of.name === 'org-access-controls').length) {
```
Can we rename "of" to something more explicit, like "feature" maybe?

raulb replied:
Changed everywhere, thanks.

naaman commented on commands/members/remove.js:
```diff
@@ -2,24 +2,69 @@
+
+    if (teamInviteFeatureEnabled) {
+      let invites = yield teamInvites()
+      isInvitedUser = !!invites.filter(m => m.user.email === email).length
```
Same here. `filter(...).length` is a weird way to check if something exists. A couple suggestions:

Array -> map -> includes:

``` javascript
> [{name: "1"},{name:"2"}].map(i => i.name).includes("2")
true
```

Array -> find

``` javascript
> [{name: "1"},{name:"2"}].find(i => i.name === "2")
{ name: '2' }
```

raulb replied:
Changed, thanks.

eablack commented on commands/access/index.js:
```diff
@@ -46,7 +46,7 @@ function * run (context, heroku) {
     try {
       let admins = yield heroku.get(`/organizations/${orgName}/members`)
-      admins = _.filter(admins, { 'role': 'admin' })
+      admins = admins.filter(a => a.role === 'admin')
```
I would probably call the initial response payload "members" instead of "admins" and then when you filter it, do something like:
`const admins = members.filter(member => member.role === 'admin')`

raulb replied:
Changed, thanks

eablack commented on commands/members/remove.js:
```diff
@@ -2,24 +2,69 @@
+
+  if (orgInfo.type === 'team') {
+    let orgFeatures = yield heroku.get(`/organizations/${groupName}/features`)
+    teamInviteFeatureEnabled = !!orgFeatures.filter(f => f.name === 'team-invite-acceptance').length
```
We should check that the feature is also enabled (i.e., f.enabled === true).

eablack replied:
Also, it probably makes sense to utilize the "find" array method instead of the filter method here.

raulb replied:
Changed, thanks.

naaman commented on commands/orgs/index.js:
```diff
@@ -2,35 +2,35 @@
+  let teams = orgs.filter(o => o.type === 'team')
+
+  // Filter by teams only
+  if (context.flags.teams) {
```
minor optimization, but the `orgs.filter` could go inside this conditional

raulb replied:
I use `teams` to educate users about the new flag `--team` (below):

``` js
if ((teams.length) && (!context.flags.teams) && (!context.flags.enterprise)) {
    cli.warn(`To list your teams only you can use ${cli.color.cmd('heroku teams')}`)
  }
```

naaman commented on commands/orgs/index.js:
```diff
@@ -2,35 +2,35 @@
+    orgs = orgs.filter(o => o.type === 'enterprise')
+  }
+
+  if (context.flags.json) Utils.printGroupsJSON(orgs)
```
Please use curly brackets to denote this condition. https://blog.codecentric.de/en/2014/02/curly-braces/

raulb replied:
Not strongly opinionated about styles here. Changed it.

eablack commented on commands/members/remove.js:
```diff
@@ -2,24 +2,69 @@
+
+    if (teamInviteFeatureEnabled) {
+      let invites = yield teamInvites()
+      isInvitedUser = !!invites.filter(m => m.user.email === email).length
```
Use find here also.

raulb replied:
Changed here too :)

naaman commented:
@raulb Sorry, I didn't use the new review feature. All done here. I think my only _real_ concern is the usage of flags vs org type. All the other stuff is pretty minor.

raulb commented:
I've implemented some of the comments raised here in this [separate PR](https://github.com/heroku/heroku-orgs/pull/62). That one only implements the invitation flow.

I'll rebase this one afterwards to only implement the new heroku team commands.

raulb commented:
@eablack @naaman thanks for your review. Just changed all the relevant suggestions. Your comments remain open (I think) due to the fact that I rebased and squashed afterwards.

This is ready for another review.

naaman commented:
👍

eablack reviewed (approved):
Ship it! :shipit:
